decimal/decimal.go

```diff
@@ -0,0 +1,109 @@
+// Package decimal with support of Tarantool's decimal data type.
+//
+// Decimal data type supported in Tarantool since 2.2.
+//
+// Since: 1.6
+//
+// See also:
+//
+// * Tarantool MessagePack extensions https://www.tarantool.io/en/doc/latest/dev_guide/internals/msgpack_extensions/#the-decimal-type
+//
+// * Tarantool data model https://www.tarantool.io/en/doc/latest/book/box/data_model/
+//
+// * Tarantool issue for support decimal type https://github.com/tarantool/tarantool/issues/692
+//
+// * Tarantool module decimal https://www.tarantool.io/en/doc/latest/reference/reference_lua/decimal/
+package decimal
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/shopspring/decimal"
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// Decimal numbers have 38 digits of precision, that is, the total
+// number of digits before and after the decimal point can be 38.
+// A decimal operation will fail if overflow happens (when a number is
+// greater than 10^38 - 1 or less than -10^38 - 1).
+//
+// See also:
+//
+// * Tarantool module decimal https://www.tarantool.io/en/doc/latest/reference/reference_lua/decimal/
+const DecimalPrecision = 38
+
+// Decimal external type.
+const decimalExtID = 1
+
+func encodeDecimal(e *msgpack.Encoder, v reflect.Value) error {
+	number := v.Interface().(decimal.Decimal)
+	one := decimal.NewFromInt(1)
+	maxSupportedDecimal := decimal.New(1, DecimalPrecision).Sub(one) // 10^DecimalPrecision - 1
```

Comment:
We do this allocation on each encode? It seems like it's not really hard to extract it to the outside.